homeassistant/components/crownstone/light.py

```diff
@@ -0,0 +1,204 @@
+
+    @property
+    def usb_available(self) -> bool:
+        """Return if this entity can use a usb dongle."""
+        return self.usb is not None and self.usb.is_ready()
+
+    @property
+    def brightness(self) -> int | None:
+        """Return the brightness if dimming enabled."""
+        return crownstone_state_to_hass(self.device.state)
+
+    @property
+    def is_on(self) -> bool:
+        """Return if the device is on."""
+        return crownstone_state_to_hass(self.device.state) > 0
+
+    @property
+    def supported_features(self) -> int:
+        """Return the supported features of this Crownstone."""
+        if self.device.abilities.get(DIMMING_ABILITY).is_enabled:
+            return SUPPORT_BRIGHTNESS
+        return 0
+
+    @property
+    def extra_state_attributes(self) -> Mapping[str, Any] | None:
+        """State attributes for Crownstone devices."""
+        attributes: dict[str, Any] = {}
+        # switch method
+        if self.usb_available:
+            attributes["switch_method"] = "Crownstone USB Dongle"
+        else:
+            attributes["switch_method"] = "Crownstone Cloud"
+
+        # crownstone abilities
+        attributes["dimming"] = ABILITY_STATE.get(
+            self.device.abilities.get(DIMMING_ABILITY).is_enabled
+        )
+        attributes["tap_to_toggle"] = ABILITY_STATE.get(
+            self.device.abilities.get(TAP_TO_TOGGLE_ABILITY).is_enabled
+        )
+        attributes["switchcraft"] = ABILITY_STATE.get(
+            self.device.abilities.get(SWITCHCRAFT_ABILITY).is_enabled
+        )
+
+        return attributes
+
+    async def async_added_to_hass(self) -> None:
+        """Set up a listener when this entity is added to HA."""
+        # new state received
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIG_CROWNSTONE_STATE_UPDATE, self.async_write_ha_state
+            )
+        )
+        # updates state attributes when usb connects/disconnects
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, SIG_UART_STATE_CHANGE, self.async_write_ha_state
+            )
+        )
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn on this light via dongle or cloud."""
+        if ATTR_BRIGHTNESS in kwargs:
+            if self.usb_available:
+                await self.hass.async_add_executor_job(
+                    partial(
+                        self.usb.dim_crownstone,
+                        self.device.unique_id,
+                        hass_to_crownstone_state(kwargs[ATTR_BRIGHTNESS]),
+                    )
+                )
+            else:
+                try:
+                    await self.device.async_set_brightness(
+                        hass_to_crownstone_state(kwargs[ATTR_BRIGHTNESS])
+                    )
+                except CrownstoneAbilityError as ability_error:
+                    _LOGGER.error(ability_error)
```

Comment:
Our current standard is to raise exceptions for errors in service handlers. The exception argument will be shown in the frontend to the user.

Reply:
Does this suggest to remove the exception here, or just the logging of it?

Reply:
Our standard is to raise a ValueError if the user has provided wrong data parameters that we can't validate in the service schema, and otherwise raise HomeAssistantError. Pass the message as exception argument. The message will be shown in the frontend.

Yes, let's remove the log message and raise an exception instead.